### PR TITLE
Make the parsers stricter, and improve unit test coverage of `mydump` package

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Improve unit-test coverage of `mydump` from 37.71% to 73.71%.

### What is changed and how it works?

We mainly lack coverage of edge cases generated by the parser. This PR introduces test cases to cover all of those. A fuzzer was used to generate the cases.

Unfortunately, the coverage of those files can only reach ~50% even with the fuzzer. That is enough to bring up total coverage of the package anyway.

The SQL parser is modified to return Syntax Error in more cases, in order to reject non-sense input like `('incomplete', 'row,`.

### Check List 

Tests 

 - Unit test

Side effects

 - Breaking backward compatibility
    * Some problematic CSV settings are now rejected (e.g. using the same character as separator and delimiter).
    * Some non-quite-well-formed SQL was previously accepted, and is now rejected.

Related changes

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @lonng @GregoryIan

File: lightning/mydump/csv_parser_test.go [hunk 1]
```diff
@@ -41,6 +41,33 @@ func (checker *assertPosEq) Check(params []interface{}, names []string) (result
...
+		c.Assert(errors.Cause(parser.ReadRow()), Equals, io.EOF, Commentf("input = %q", tc.input))
+	}
+}
+
+func (s *testMydumpCSVParserSuite) runFailingTestCases(c *C, cfg *config.CSVConfig, blockBufSize int64, cases []string) {
+	for _, tc := range cases {
+		parser := mydump.NewCSVParser(cfg, strings.NewReader(tc), blockBufSize, s.ioWorkers)
+		c.Assert(parser.ReadRow(), ErrorMatches, "Syntax error", Commentf("input = %q", tc))
+	}
+
```
Comment (lonng) [inline lightning/mydump/csv_parser_test.go]: Remove the empty line.

File: lightning/mydump/csv_parser_test.go [hunk 2]
```diff
@@ -383,3 +418,717 @@ func (s *testMydumpCSVParserSuite) TestEmpty(c *C) {
...
+		{
+			input:    ",0",
+			expected: [][]types.Datum{{nullDatum, types.NewStringDatum("0")}},
+		},
+		{
+			input:    ",\r",
+			expected: [][]types.Datum{{nullDatum, nullDatum}},
+		},
+		{
+			input:    ",\n",
```
Comment (lonng) [inline lightning/mydump/csv_parser_test.go]: add case for `,\r\n`

File: lightning/mydump/parser.go [hunk 3]
```diff
@@ -371,6 +371,9 @@ func (parser *ChunkParser) ReadRow() error {
 	for {
 		tok, content, err := parser.lex()
 		if err != nil {
+			if err == io.EOF && st != stateValues {
+				return errors.Errorf("Syntax error: premature EOF at offset %d", parser.pos)
```
Comment (lonng) [inline lightning/mydump/parser.go]: I think the error message starting with lower-case is more appropriate. See: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: lightning/mydump/parser.go [hunk 4]
```diff
@@ -380,6 +383,12 @@ func (parser *ChunkParser) ReadRow() error {
 				st = stateColumns
 			case tokValues:
 				st = stateValues
+			case tokUnquoted, tokDoubleQuoted, tokBackQuoted:
+			default:
+				return errors.Errorf(
+					"Syntax error: unexpected %s (%s) at offset %d, expecting %s",
```
Comment (lonng) [inline lightning/mydump/parser.go]: (ditto)

File: lightning/mydump/parser.go [hunk 5]
```diff
@@ -388,6 +397,11 @@ func (parser *ChunkParser) ReadRow() error {
 			case tokUnquoted, tokDoubleQuoted, tokBackQuoted:
 				columnName := parser.unescapeString(string(content))
 				parser.columns = append(parser.columns, columnName)
+			default:
+				return errors.Errorf(
+					"Syntax error: unexpected %s (%s) at offset %d, expecting %s",
```
Comment (lonng) [inline lightning/mydump/parser.go]: (ditto)

File: lightning/mydump/parser.go [hunk 6]
```diff
@@ -398,6 +412,12 @@ func (parser *ChunkParser) ReadRow() error {
 			case tokUnquoted, tokDoubleQuoted, tokBackQuoted:
 				parser.columns = nil
 				st = stateTableName
+			case tokValues:
+			default:
+				return errors.Errorf(
+					"Syntax error: unexpected %s (%s) at offset %d, expecting %s",
```
Comment (lonng) [inline lightning/mydump/parser.go]: (ditto)

File: lightning/mydump/parser.go [hunk 7]
```diff
@@ -444,7 +464,10 @@ func (parser *ChunkParser) ReadRow() error {
 				}
 				value.SetBinaryLiteral(binLit)
 			default:
-				return errors.Errorf("Syntax error at position %d", parser.pos)
+				return errors.Errorf(
+					"Syntax error: unexpected %s (%s) at offset %d, expecting %s",
```
Comment (lonng) [inline lightning/mydump/parser.go]: (ditto)

File: lightning/mydump/parser.rl [hunk 8]
```diff
@@ -172,7 +171,11 @@ func (parser *ChunkParser) lex() (token, []byte, error) {
 		}
 
 		if parser.isLastChunk {
-			return tokNil, nil, io.EOF
+			if te == eof {
+				return tokNil, nil, io.EOF
+			} else {
+				return tokNil, nil, errors.New("Syntax error: unexpected EOF")
```
Comment (lonng) [inline lightning/mydump/parser.rl]: (ditto)

**Comment by kennytm** [comment]
Please check again @lonng

**Comment by kennytm** [comment]
PTAL @july2993

File: Makefile [hunk 9]
```diff
@@ -24,7 +24,7 @@ GOTEST    := GO111MODULE=on CGO_ENABLED=1 $(GO) test -p 3
 ARCH      := "`uname -s`"
 LINUX     := "Linux"
 MAC       := "Darwin"
-PACKAGES  := $$(go list ./...| grep -vE 'vendor|cmd|test|proto|diff|bin')
+PACKAGES  := $$(go list ./...| grep -vE 'vendor|cmd|test|proto|diff|bin|fuzz')
```
Comment (july2993) [inline Makefile]: where's the fuzz pkg?

Comment (kennytm) [reply Makefile]: @july2993 Not included here. I'm thinking how to introduce fuzz test, whether directly here or as a different repository, since it has amassed over 3000 corpus already.
